Luminance-based window shading control per 90.1 Appendix C requirements

leijerry888 opened the pull request:
Pull request overview
---------------------
 - Addressing Issue #7081
 - Adding three window shading control options that turn on window shading based on luminance or solar gain threshold (90.1 2019 Appendix C Section C3.5.5.1), and turn off shading at different time / sun conditions.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [x] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [x] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [x] Check any new function arguments for performance impacts
 - [x] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

leijerry888 committed:
nfp draft

leijerry888 committed:
update NFP document

leijerry888 committed:
Merge branch 'develop' into nfp_901appc_shading

leijerry888 committed:
update NFP and add luminance variable [no ci]

leijerry888 committed:
update NFP [decent_ci_skip] [actions skip]

leijerry888 committed:
update NFP [decent_ci_skip] [actions skip]

leijerry888 committed:
update NFP [decent_ci_skip] [actions skip]

leijerry888 committed:
revert changes of adding incorrect luminance variable

leijerry888 committed:
Merge branch 'develop' into nfp_901appc_shading

leijerry888 committed:
add core logic for luminance based shading control (WIP)

leijerry888 committed:
bug fix and clean up

leijerry888 committed:
Add two more window shading control options

leijerry888 committed:
response to comment, replace hard-coded array size

leijerry888 committed:
Merge branch 'develop' into nfp_901appc_shading

leijerry888 committed:
update code after reverse merge

leijerry888 committed:
design document cleanup

leijerry888 committed:
Merge branch 'develop' into nfp_901appc_shading

leijerry888 committed:
add luminance shading control control unit test

leijerry888 committed:
add window shading control type unit test

leijerry888 committed:
clang format

leijerry888 committed:
Merge branch 'develop' into nfp_901appc_shading

leijerry888 committed:
update i/o ref

leijerry888 committed:
update idd note

leijerry888 committed:
update design document [decent_ci_skip]

leijerry888 committed:
respond to comments

leijerry888 commented on src/EnergyPlus/DataEnvironment.hh:
```diff
@@ -135,6 +135,7 @@ struct EnvironmentData : BaseGlobalStruct
     Real64 SkyTempKelvin = 0.0;                   // Current sky temperature {K}
     Real64 LiquidPrecipitation = 0.0;             // Current liquid precipitation amount (rain) {m}
     bool SunIsUp = false;                         // True when Sun is over horizon, False when not
+    bool SunIsUpPrevTS = false;                   // True when Sun is over horizon in the previous timestep, False when not in the previous timestep
```
adding variable for sun status of the previous time step for use of checking "next morning"

leijerry888 commented on src/EnergyPlus/DataSurfaces.hh:
```diff
@@ -153,28 +153,31 @@ namespace DataSurfaces {
+        HiLumin_HiSolar_OffMidNight,
+        HiLumin_HiSolar_OffSunset,
+        HiLumin_HiSolar_OffNextMorning,
         Num
```
add new control options

leijerry888 commented on src/EnergyPlus/DataSurfaces.hh:
```diff
@@ -348,29 +351,6 @@ namespace DataSurfaces {
-    constexpr int WSCT_OnHiOutTemp_HiHorzSolar(19);       // OnIfHighOutsideAirTempAndHighHorizontalSolar
-    constexpr int WSCT_OnHiZoneTemp_HiSolarWindow(20);    // OnIfHighZoneAirTempAndHighSolarOnWindow
-    constexpr int WSCT_OnHiZoneTemp_HiHorzSolar(21);      // OnIfHighZoneAirTempAndHighHorizontalSolar
-
```
delete because they are not used anywhere

leijerry888 commented on src/EnergyPlus/DaylightingManager.cc:
```diff
@@ -6815,6 +6815,46 @@ void DayltgInteriorIllum(EnergyPlusData &state,
+            }
+        }
+    }
+
```
luminance checking and shading control

leijerry888 commented on src/EnergyPlus/SolarShading.cc:
```diff
@@ -9736,6 +9736,69 @@ void WindowShadingManager(EnergyPlusData &state)
+                    shadingOffButGlareControlOn = false;
+                }
+                break;
+
```
three new window shading control options and control logic (except for luminance control part)

leijerry888 commented on src/EnergyPlus/SurfaceGeometry.cc:
```diff
@@ -9615,7 +9615,10 @@ namespace SurfaceGeometry {
+                                                                     "ONIFHIGHZONEAIRTEMPANDHIGHHORIZONTALSOLAR",
+                                                                     "ONIFHIGHLUMINANCEORHIGHSOLARTILLMIDNIGHT",
+                                                                     "ONIFHIGHLUMINANCEORHIGHSOLARTILLSUNSET",
+                                                                     "ONIFHIGHLUMINANCEORHIGHSOLARTILLNEXTMORNING"});
```
adding new shading control options

amirroth replied:
We are replacing these `Array1D_string` with `constexpr std::array<std::string_view, static_cast<int>(X::Num)> XNamesUC = {};`. These arrays are compile time structures that are just loaded from the executable data section rather than constructed on the heap.

amirroth replied:
While you are at this, you will have to replace the corresponding `FindItemInList` with `getEnumerationValue`, which has the equivalent functionality but also works with these compile-time constant arrays. There are examples of this throughout the code.

leijerry888 replied:
Thanks! getEnumerationValue fixed the problem.

mjwitte commented on src/EnergyPlus/DaylightingManager.cc:
```diff
@@ -6815,6 +6815,46 @@ void DayltgInteriorIllum(EnergyPlusData &state,
+    // loop over windows to do luminance based control
+    int count = 0;
+    for (std::size_t igroup = 1; igroup <= thisDaylightControl.ShadeDeployOrderExtWins.size(); igroup++) {
+        std::vector<int> const &listOfExtWin = thisDaylightControl.ShadeDeployOrderExtWins[igroup - 1];
+        for (const auto IWin : listOfExtWin) {
```
This reference is used only once in the next line.

amirroth replied:
`IWin` is not an `auto`, it's an `int`.

leijerry888 replied:
Comments addressed. I was following similar coding patterns in the same file.

amirroth replied:
Not a generally good thing to do in EnergyPlus right now. Unless you’re in a module that has been swept thoroughly within the last six months to a year (and there are several), you will be perpetuating a lot of old idioms.

leijerry888 replied:
> Not a generally good thing to do in EnergyPlus right now. Unless you’re in a module that has been swept thoroughly within the last six months to a year (and there are several), you will be perpetuating a lot of old idioms.

Thanks for the note!

mjwitte commented on src/EnergyPlus/DaylightingManager.cc:
```diff
@@ -6815,6 +6815,46 @@ void DayltgInteriorIllum(EnergyPlusData &state,
+            int loop = thisDaylightControl.MapShdOrdToLoopNum(count);
+            if (loop > 0) {
+                ICtrl = state.dataSurface->Surface(IWin).activeWindowShadingControl;
+                if (!state.dataSurface->Surface(IWin).HasShadeControl) continue;
```
This seems unnecessary. Isn't `HasShadeControl` true for every window in `ShadeDeployOrderExtWins`?

leijerry888 replied:
I think so. Code changed accordingly. Was following similar coding patterns.

mjwitte commented on src/EnergyPlus/DaylightingManager.cc:
```diff
@@ -6815,6 +6815,46 @@ void DayltgInteriorIllum(EnergyPlusData &state,
+                WindowShadingControlType shCtrlType = state.dataSurface->WindowShadingControl(ICtrl).ShadingControlType;
+                if (((shCtrlType == WindowShadingControlType::HiLumin_HiSolar_OffMidNight) ||
+                     (shCtrlType == WindowShadingControlType::HiLumin_HiSolar_OffSunset) ||
+                     (shCtrlType == WindowShadingControlType::HiLumin_HiSolar_OffNextMorning)) &&
```
Since this entire loop applies only to these three control types, check this first with `if (! (a && b && c)) continue;` ?

leijerry888 replied:
Great point. Comment addressed.

mjwitte commented on src/EnergyPlus/DaylightingManager.cc:
```diff
@@ -6815,6 +6815,46 @@ void DayltgInteriorIllum(EnergyPlusData &state,
+                if (((shCtrlType == WindowShadingControlType::HiLumin_HiSolar_OffMidNight) ||
+                     (shCtrlType == WindowShadingControlType::HiLumin_HiSolar_OffSunset) ||
+                     (shCtrlType == WindowShadingControlType::HiLumin_HiSolar_OffNextMorning)) &&
+                    ((currentFlag == WinShadingType::IntShadeConditionallyOff) || (currentFlag == WinShadingType::GlassConditionallyLightened) ||
+                     (currentFlag == WinShadingType::ExtShadeConditionallyOff) || (currentFlag == WinShadingType::IntBlindConditionallyOff) ||
+                     (currentFlag == WinShadingType::ExtBlindConditionallyOff) || (currentFlag == WinShadingType::BGShadeConditionallyOff) ||
+                     (currentFlag == WinShadingType::BGBlindConditionallyOff))) {
```
This looks like a BITF candidate? @amirroth ?

amirroth replied:
Ummm ... so we don't have to do this anymore I don't think. Someone please check MSVC but g++ 4.2 does the `BITF` equivalent optimization. Even under `-O`!

mjwitte replied:
> Ummm ... so we don't have to do this anymore I don't think. Someone please check MSVC but g++ 4.2 does the `BITF` equivalent optimization. Even under `-O`!

Oh, ok, didn't know that.

amirroth replied:
So, whatever g++ version we were on three years ago did not do this. I looked again a few weeks ago because of something Yujie asked me and was pleasantly surprised! Someone please check what the situation is with MSVC.

mjwitte replied:
> So, whatever g++ version we were on three years ago did not do this. I looked again a few weeks ago because of something Yujie asked me and was pleasantly surprised! Someone please check what the situation is with MSVC.

How would I check that in MSVC?

mjwitte commented on src/EnergyPlus/SolarShading.cc:
```diff
@@ -9736,6 +9736,69 @@ void WindowShadingManager(EnergyPlusData &state)
+                        shadingOffButGlareControlOn = true;
+                    }
+                }
+                // if it is the beginning of the day, then shades off
+                if ((state.dataGlobal->HourOfDay - 1) * state.dataGlobal->NumOfTimeStepInHour + state.dataGlobal->TimeStep == 1) {
```
`if (state.dataGlobal->BeginDayFlag)`

leijerry888 replied:
Code changed accordingly

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -12801,11 +12809,14 @@ WindowShadingControl,
   N3 , \field Setpoint 2
-       \units W/m2 or deg C
-       \note W/m2 for solar-based controls, deg C for temperature-based controls.
+       \units W/m2, deg C, or cd/m2
```
IDF Editor does not recognize this units string and it will throw a warning every time it is opened for any idf file. @JasonGlazer Do we want to update IDF Editor at this point for a set of units that won't get converted? Or just remove the units field here?

JasonGlazer replied:
@mjwitte it is very simple to add units to the IDF Editor so I think we should try to do that.

mjwitte reviewed (commented):
Some initial comments.
A new example file which uses these options would be useful. Perhaps copy an existing windowshading control file and add these options?

mjwitte commented:
I made a quick testfile,[WindowTests-UseNewControls.idf.txt](https://github.com/NREL/EnergyPlus/files/9351903/WindowTests-UseNewControls.idf.txt) to test these new shading control options. It fails in a debug build [here](https://github.com/NREL/EnergyPlus/blob/f7a5dbf0db4584d27c83d33fd38e6702ed603a80/src/EnergyPlus/SolarShading.cc#L11999) because `ShBlDifSolarAbsW` is uninitialized. This isn't code you changed, but perhaps something is getting missed in the ifs above? Or maybe this just needs an initialization somewhere up above. Not sure what the correct fix is.

amirroth commented on tst/EnergyPlus/unit/SolarShading.unit.cc:
```diff
@@ -2655,6 +2657,84 @@ TEST_F(EnergyPlusFixture, SolarShadingTest_selectActiveWindowShadingControl)
+    surf2.activeShadedConstruction = 1;
+
+    state->dataConstruction->Construct(1).Name = "Construction1";
+
```
May also want to use `auto &windowShadingControl1;` here.

leijerry888 committed:
address src PR comments

leijerry888 committed:
address tst PR comment

leijerry888 committed:
clang format

leijerry888 commented:
> Some initial comments. A new example file which uses these options would be useful. Perhaps copy an existing windowshading control file and add these options? 

I modified the WindowTests example idf file by replacing shading control of two windows ( from the four windows under MeetDaylightIlluminanceSetpoint control) with a new luminance based shading control option and it works. It is not in the PR at this moment. I will push it for review.

leijerry888 commented on doc/input-output-reference/src/overview/group-thermal-zone-description-geometry.tex:
```diff
@@ -4464,6 +4464,15 @@ \subsubsection{Inputs}\label{inputs-32-001}
 
+The following three control types can be used to model human behavior of turning on shading during the day to reduce view luminance or solar gain and leaving shading on until a certain time point. They can only be used in zones with daylighting controls. Window view luminance at the first daylighting reference point is used to compare with luminance setpoint. They are applicable to any Shading Ty ...
+
+\emph{OnIfHighLuminanceOrHighSolarTillMidnight:} Shading is on if the solar radiation incident on the window exceeds SetPoint (W/m\(^{2}\)) or the luminance from the window exceeds SetPoint 2 (cd/m\(^{2}\)). When shading is on, it remains on till midnight.
```
@mjwitte and others, a question about naming and setpoint order contention: currently the new options I added are named "On if A Or B Till XX", and they all use two setpoints under WindowShadingControl, where setpoint is for B and setpoint2 is for A. i.e. the sequence of A and B are not aligned between option name and setpoint order. Is this okay or do we want to make sure they align, by either changing naming or changing setpoint sequence?

mjwitte replied:
These controls have become so complicated, I'm not sure if the order of the setpoints matter, but I suppose it would be best to align them, so I would vote for changing the control type names.

leijerry888 replied:
Thanks. Names aligned.

leijerry888 committed:
align control option name with setpoint order

leijerry888 committed:
modify window example file to include new control option

leijerry888 committed:
minor format change to align style in idd

leijerry888 commented:
> > Some initial comments. A new example file which uses these options would be useful. Perhaps copy an existing windowshading control file and add these options?
> 
> I modified the WindowTests example idf file by replacing shading control of two windows ( from the four windows under MeetDaylightIlluminanceSetpoint control) with a new luminance based shading control option and it works. It is not in the PR at this moment. I will push it for review.

@mjwitte modified windowtests.idf committed.

leijerry888 committed:
Merge branch 'develop' into nfp_901appc_shading

leijerry888 committed:
update setpoint2 units in non test files

leijerry888 committed:
update setpoint2 units in testfiles

leijerry888 committed:
update design document final

leijerry888 commented on src/IDF_Editor/IDFEditMain.bas:
```diff
@@ -694,7 +694,7 @@ convUnits(59).siName = "eV"
 convUnits(61).siName = "percentage (as a real decimal)"
 convUnits(62).siName = "s"
-convUnits(63).siName = "W/m2 or deg C"
+convUnits(63).siName = "W/m2, deg C or cd/m2"
```
@mjwitte @JasonGlazer I updated the units here in IDF_Editor and also updated all occurrences of 'W/m2 or deg C' to 'W/m2, deg C or cd/m2'. It seems like this setpoint2 field is the only place using it.

leijerry888 commented on src/IDF_Editor/IDFEditMain.bas:
```diff
@@ -2440,7 +2440,7 @@ End Sub
 '     the specific language governing rights and limitations under the License.
 '
-'     Copyright � 1996-2014 GARD Analytics.  All rights reserved.
+'     Copyright � 1996-2014 GARD Analytics.  All rights reserved.
```
not sure what encoding issue on my editor caused this being a change. sorry...

mjwitte reviewed (changes requested):
- Unit tests pass locally.
- Original WindowTests.idf produces same results with develop and this branch.
- Ok on the new choice names to align with setpoint order.
- @JasonGlazer Is addressing the IDFEditor units issue in a separate PR, so please revert the IDF Editor source file changes to avoid conflicts.
- My testfile still crashes as noted [here](https://github.com/NREL/EnergyPlus/pull/9571#issuecomment-1216720904). Revised version of the test file (new choice keys) is attached. [WindowTests-UseNewControls-Rev.idf.txt](https://github.com/NREL/EnergyPlus/files/9383412/WindowTests-UseNewControls-Rev.idf.txt) It's quite possible that there is an error in one of the modified inputs which needs to be trapped during getInput.

leijerry888 committed:
revert unit change in IDFEdit

leijerry888 commented:
> * Unit tests pass locally.
> * Original WindowTests.idf produces same results with develop and this branch.
> * Ok on the new choice names to align with setpoint order.
> * @JasonGlazer Is addressing the IDFEditor units issue in a separate PR, so please revert the IDF Editor source file changes to avoid conflicts.
> * My testfile still crashes as noted [here](https://github.com/NREL/EnergyPlus/pull/9571#issuecomment-1216720904). Revised version of the test file (new choice keys) is attached. [WindowTests-UseNewControls-Rev.idf.txt](https://github.com/NREL/EnergyPlus/files/9383412/WindowTests-UseNewControls-Rev.idf.txt) It's quite possible that there is an error in one of the modified inputs which needs to be trapped during getInput.

Thanks @mjwitte for the comments.
* I reverted the unit changes in IDFEditMain.bas
* I checked your example. Error seems to be caused by 2 out of the 3 WindowShadingControl objects for new options do not have DaylightingControl object associated. I added 2 pairs of DaylightingControl and referencepoint objects and they work now.

mjwitte commented:
> * I checked your example. Error seems to be caused by 2 out of the 3 WindowShadingControl objects for new options do not have DaylightingControl object associated. I added 2 pairs of DaylightingControl and referencepoint objects and they work now.

Ok, but the windowshading input processing needs to check this, throw a severe error, and set the errorsfound flag so that the simulation stops. I assume the glare control options check this, so the new options should also.

JasonGlazer commented:
@leijerry888 I put the IDF Editor fix to the units as well as a couple of other fixes in #9604. Thanks for letting me know about the new units.

leijerry888 committed:
add check for daylighting control object name

leijerry888 commented:
> 



> > * I checked your example. Error seems to be caused by 2 out of the 3 WindowShadingControl objects for new options do not have DaylightingControl object associated. I added 2 pairs of DaylightingControl and referencepoint objects and they work now.
> 
> Ok, but the windowshading input processing needs to check this, throw a severe error, and set the errorsfound flag so that the simulation stops. I assume the glare control options check this, so the new options should also.

The check is added https://github.com/NREL/EnergyPlus/pull/9571/commits/b9e70e3aaa39a06713fea3c42bf4753dc189f6e8

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into nfp_901appc_shading

mjwitte reviewed (approved):
Merged in develop.
Unit tests pass locally.
Test file generates error as expected now.
Will let CI run and then will merge.
Thanks @leijerry888

mjwitte commented:
@Myoldmopar @jmythms When this merges later today it will cause conflicts with #9575

jmythms commented:
> @Myoldmopar @jmythms When this merges later today it will cause conflicts with #9575 

No worries. I will try to update #9575 after this merges. Will ask for help if I run into trouble.

Myoldmopar commented:
It's too bad GitHub stopped showing commit messages on PRs (https://github.blog/changelog/2022-08-04-commit-comments-no-longer-appear-in-the-pull-request-timeline/).  We'll need to figure out how to handle that in our CI.  

Anyway, if you look at the dashboard, there appears to be a minor build warning to go along with the WindowTests diffs: https://github.blog/changelog/2022-08-04-commit-comments-no-longer-appear-in-the-pull-request-timeline/.  If the build warning is the only thing keeping this from going in, I'm happy to remove that line and push/merge.

Thanks for the work @leijerry888  and the review @mjwitte

Myoldmopar commented:
Oh, sorry, noticed that build warning already slipped into develop, so it's nothing to do with this, merge when you feel it's ready @mjwitte

leijerry888 commented:
Thanks @amirroth @mjwitte @Myoldmopar @JasonGlazer for the review and help!
